bindings/mysql/mysql.go

```diff
@@ -0,0 +1,281 @@
+			return err
+		}
+		m.db = db
+	} else {
+		db, err := initDBFromConfig(p[userKey], p[passwordKey], p[networkKey], p[serverKey], p[portKey], p[databaseKey])
+		if err != nil {
+			return err
+		}
+		m.db = db
+	}
+
+	if err := propertyToInt(p, maxIdleConnsKey, m.db.SetMaxIdleConns); err != nil {
+		return err
+	}
+
+	if err := propertyToInt(p, maxOpenConnsKey, m.db.SetMaxOpenConns); err != nil {
+		return err
+	}
+
+	if err := propertyToDuration(p, connMaxIdleTimeKey, m.db.SetConnMaxIdleTime); err != nil {
+		return err
+	}
+
+	if err := propertyToDuration(p, connMaxLifetimeKey, m.db.SetConnMaxLifetime); err != nil {
+		return err
+	}
+
+	if err := m.db.Ping(); err != nil {
+		return errors.Wrap(err, "unable to pint the DB")
+	}
+
+	return nil
+}
+
+// Invoke handles all invoke operations
+func (m *Mysql) Invoke(req *bindings.InvokeRequest) (*bindings.InvokeResponse, error) {
+	if req == nil {
+		return nil, errors.Errorf("invoke request required")
+	}
+
+	if req.Operation == closeOperation {
+		return nil, m.db.Close()
+	}
+
+	if req.Metadata == nil {
+		return nil, errors.Errorf("metadata required")
+	}
+	m.logger.Debugf("operation: %v", req.Operation)
+
+	s, ok := req.Metadata[commandSQLKey]
+	if !ok || s == "" {
+		return nil, errors.Errorf("required metadata not set: %s", commandSQLKey)
+	}
+
+	startTime := time.Now().UTC()
+
+	resp := &bindings.InvokeResponse{
+		Metadata: map[string]string{
+			respOpKey:        string(req.Operation),
+			respSQLKey:       s,
+			respStartTimeKey: startTime.Format(time.RFC3339Nano),
+		},
+	}
+
+	switch req.Operation { // nolint: exhaustive
+	case execOperation:
+		r, err := m.exec(s)
+		if err != nil {
+			return nil, errors.Wrapf(err, "error executing %s with %v", s, err)
+		}
+		resp.Metadata[respRowsAffectedKey] = strconv.FormatInt(r, 10)
+
+	case queryOperation:
+		d, err := m.query(s)
+		if err != nil {
+			return nil, errors.Wrapf(err, "error executing %s with %v", s, err)
+		}
+		resp.Data = d
+
+	default:
+		return nil, errors.Errorf("invalid operation type: %s. Expected %s, %s, or %s",
+			req.Operation, execOperation, queryOperation, closeOperation)
+	}
+
+	endTime := time.Now().UTC()
+	resp.Metadata[respEndTimeKey] = endTime.Format(time.RFC3339Nano)
+	resp.Metadata[respDurationKey] = endTime.Sub(startTime).String()
+
+	return resp, nil
+}
+
+// Operations returns list of operations supported by Mysql binding
+func (m *Mysql) Operations() []bindings.OperationKind {
+	return []bindings.OperationKind{
+		execOperation,
+		queryOperation,
+		closeOperation,
+	}
+}
+
+func (m *Mysql) query(s string) ([]byte, error) {
+	m.logger.Debugf("query: %s", s)
+
+	rows, err := m.db.Query(s)
+	if err != nil {
+		return nil, errors.Wrapf(err, "error executing %s", s)
+	}
+
+	defer func() {
+		_ = rows.Close()
+		_ = rows.Err()
+	}()
+
+	columns, err := rows.Columns()
+	if err != nil {
+		return nil, errors.Wrapf(err, "error finding column names for %s", s)
+	}
+
+	var values [][]sql.RawBytes
```

Comment:
Hi @beiwei30!  How did you come up with using `sql.RawBytes` for this?  Its very concise but this documentation for it makes me wonder if it is safe to use like this.

```
// RawBytes is a byte slice that holds a reference to memory owned by
// the database itself. After a Scan into a RawBytes, the slice is only
// valid until the next call to Next, Scan, or Close.
```

There were other approaches that I found here. Could any of these work?

https://stackoverflow.com/questions/42774467/how-to-convert-sql-rows-to-typed-json-in-golang

Reply:
I simply follow the sample code of 'RawBytes' from mysql's doc: https://github.com/go-sql-driver/mysql/wiki/Examples#rawbytes

In fact, RawBytes is simply a slice of byte:

```go
type RawBytes []byte
``` 

In my impl, I create sql.RawBytes for each iteration instead of reuse it so that the behavior described in the comments quoted above is not a concern.

Reply:
Thanks for the example link and agreed on the safety by creating a new row of sql.RawBytes for each iteration.

I'm also curious how numeric types are marshaled in the JSON...  does 1234 serialize to `1234` or `"1234"` (with quotes)?

Reply:
I will try json marshal tomorrow and get back to you later.

Reply:
@pkedy would you mind to take a look into https://github.com/dapr/components-contrib/pull/615/commits/dab941ece703d31c489c7b27c76b9878e8a22487#diff-6b618681e6b9a3154c568fe120e86733654e2f369edfea503df583ac5c3e8ef1R197

Reply:
> Thanks for the example link and agreed on the safety by creating a new row of sql.RawBytes for each iteration.
> 
> I'm also curious how numeric types are marshaled in the JSON... does 1234 serialize to `1234` or `"1234"` (with quotes)?

pls. check https://github.com/dapr/components-contrib/pull/615/files#diff-dc91a6eb8044921dc5f73b0a7dd0e22fd2b03e619862931032b95fb2e9965cdfR48. For numeric types, it's marshalled into a number instead of a string.

Reply:
@beiwei30 I see that test :+1: and I ❤️ the changes you've made since I last looked at this. You might also consider adding an `assert.Contains` for the `TIME` value you have (for example does that serialize as a RFC 3339 string or integer in milliseconds) and maybe adding a `boolean` type to make sure that comes back as `true` or `false` and not `1` or `0`. 

What do you think about adding/calling your type tests from the integration tests also?  My thought is that `go-sqlmock` could return stuff differently than the MySQL client. Just a thought.

Reply:
@pkedy pls. check  61c286d08a38df409479789415bea38781eeba87 and b6ce23db57e3df1d7eac131fac4caef36baef79c

Reply:
with regarding to `boolean` type, in fact, mysql doesn't support `BOOL` at all.  I add a test to verify it is `TINYINT(1)` anyway. Pls. check this 62572d9e45a8de96b88ca7111f43f9eeab512736